### PR TITLE
Python virtualenv warning

### LINKED ISSUE #4471: syslog-ng -s prints warning, if Python support is installed but venv not initialized
Version 4.2.0 of syslog-ng shows a warning message when ```syslog-ng -s``` is run, if Python support is installed, but virtualenv is not initialized. My RPM packages use Python dependencies from the system Python installation:
```
czplaptop:~ # syslog-ng -s
[2023-05-12T12:25:00.518004] python: private virtualenv is not initialized, use the `syslog-ng-update-virtualenv' script to initialize it or make sure all required Python dependencies are available in the system Python installation; path='/var/lib/syslog-ng/python-venv'
```
The problem is, that quite a few users run ```syslog-ng -s``` in scripts before restarting syslog-ng, and those fail if there is any output from syslog-ng.

### PR DESCRIPTION
**Pull request by bazsi**
This should fix #4471

### CODE REVIEW

File: modules/python/python-main.c [hunk 1]
```diff
@@ -305,14 +305,13 @@ _get_venv_path(void)
...
 
-
   const gchar *private_venv_path = get_installation_path_for(SYSLOG_NG_PYTHON_VENV_DIR);
   if (!_py_is_virtualenv_valid(private_venv_path))
     {
-      msg_warning("python: private virtualenv is not initialized, use the `syslog-ng-update-virtualenv' "
-                  "script to initialize it or make sure all required Python dependencies are available in "
-                  "the system Python installation",
-                  evt_tag_str("path", private_venv_path));
+      msg_debug("python: private virtualenv is not initialized, use the `syslog-ng-update-virtualenv' "
```
Comment (MrAnno) [inline modules/python/python-main.c:311]: I don't understand this change (why we need to move this to the debug level).

If our virtualenv is optional, that should be reflected in the value of the `use_virtualenv` flag in my opinion.

None of the virtualenv-related error messages are shown when `use_virtualenv` is false, which currently requires the user (or the syslog-ng package maintainer) to set `@module mod-python use-virtualenv(no)`.

If setting `@module mod-python use-virtualenv(no)` is problematic, we could provide another way to disable the virtualenv (compile time flag, or something that can be disabled during packaging). My understanding is that the user (or package maintainer) decides if they want to use system Python dependencies or the venv, and once that decision is made, syslog-ng should provide error messages if something isn't right.

Maybe I'm wrong and we just want to fall back silently, but in that case, the `use_virtualenv` flag seems to be unnecessary.

File: lib/ivykis
Comment (MrAnno) [inline lib/ivykis:1]: There is an ivykis change in this commit that is probably accidental.

**Comment by bazsi** [comment]
I simply forgot that we have that option (even though I probably
implemented it).

The issue is with existing configs, which obviously doesn't specify
use-virtualenv option. If we would want to keep that, then we should have
defaulted to not using a virtualenv unless explicitly enabled.

But I believe that in most cases using a virtualenv is the better choice,
except for distro packages, which want to use python dependencies they
package themselves.

But distro packages will not want to force the users to use the
use-virtualenv command to python.

Let's discuss this IRL or discord.

**Comment by czanik** [comment]
Yeah, @bazsi is right. It is about distro packaging, and not wanting to touch configurations. A ```configure``` option would be fantastic. That way I could turn off the warning when Python dependencies are available in the distro.

Thinking a bit more, there is yet another corner case. Right now I'm working on splitting the drivers written in Python (Kubernetes, Hyper) from the syslog-ng Python module. On a distro where Python dependencies are unavailable:

 * if the Python-based drivers are installed then we definitely need virtualenv
 * if the Python-based drivers are not installed, there is a good chance that the user can develop something using purely distro-based Python dependencies, so the warning is useless

And one more scenario came to my mind: what if the user does not install our Python-based drivers, but still needs virtualenv? Right now the ```syslog-ng-update-virtualenv``` has the dependency list hard coded:

```
REQUIREMENTS_FILE=${python_moduledir}/requirements.txt
```

File: modules/python/python-main.c (hunk 1 above)
Comment (bazsi) [reply modules/python/python-main.c:311]: What do you think about the current solution? I just pushed. Moved all messages to debug at the same time every time we get a ModuleNotFoundError we report that the user might need to create the virtualenv.

Cheers,
Bazsi

**Comment by bazsi** [comment]
> Yeah, @bazsi is right. It is about distro packaging, and not wanting to touch configurations. A `configure` option would be fantastic. That way I could turn off the warning when Python dependencies are available in the distro.

Can you check the current solution? I've moved all warnings/info messages to debug  level. syslog-ng will try to activate the virtualenv if it's available but remains silent if it's not.

Then we report this if we get a ModuleNotFoundError:

      msg_error("Seems you are missing a module that may be referenced by a "
                "syslog-ng plugin implemented in Python. These modules "
                "need to be installed either using your platform's package management "
                "tools (e.g. apt/dnf/yum) or Python's own package management "
                "tool (e.g. pip). syslog-ng authors recommend using pip and "
                "a dedicated Python virtualenv. You can initialize such a "
                "virtualenv using the `syslog-ng-update-virtualenv` command. "
                "This command will initialize the virtualenv and install all "
                "packages needed by plugins shipped with syslog-ng itself "
                "from the Python Package Index (PyPI). If you need any additional "
                "Python libraries for your local scripts, you can "
                "install those using the `pip` command located in the virtualenv's bin directory",
                evt_tag_str("module", module_cstr));


> 
> Thinking a bit more, there is yet another corner case. Right now I'm working on splitting the drivers written in Python (Kubernetes, Hyper) from the syslog-ng Python module. On a distro where Python dependencies are unavailable:
> 
>     * if the Python-based drivers are installed then we definitely need virtualenv
> 
>     * if the Python-based drivers are not installed, there is a good chance that the user can develop something using purely distro-based Python dependencies, so the warning is useless
> 
> 
> And one more scenario came to my mind: what if the user does not install our Python-based drivers, but still needs virtualenv? Right now the `syslog-ng-update-virtualenv` has the dependency list hard coded:
> 
> ```
> REQUIREMENTS_FILE=${python_moduledir}/requirements.txt
> ```

you can always run pip from inside the virtualenv to add more packages. This is just bootstrapping that directory.

**Review by MrAnno — APPROVED** [review]
LGTM

**Comment by czanik** [comment]
@bazsi  Looks good to me. This week I was focusing on he BalaBit Hackathon, but once I'm back from the openSUSE conference in Nürnberg, I'll back to doing the packaging split, and I'll test these changes as well.

**Comment by czanik** [comment]
Thanks @bazsi . Tested on FreeBSD, and works as expected. When I'll have a bit more time, I'll check the ```syslog-ng-update-virtualenv``` command, how to install a custom list of dependencies, instead of 50+MB of Kubernetes-related Python code...
```
root@fb132:/var/db # rm -fr python-venv/
root@fb132:/var/db # syslog-ng -s
[2023-05-30T18:24:16.010334] Error evaluating global Python block; exception='ModuleNotFoundError: No module named \'requests\''
[2023-05-30T18:24:16.010401] Seems you are missing a module that may be referenced by a syslog-ng plugin implemented in Python. These modules need to be installed either using your platform's package management tools (e.g. apt/dnf/yum) or Python's own package management tool (e.g. pip). syslog-ng authors recommend using pip and a dedicated Python virtualenv. You can initialize such a virtualenv using the `syslog-ng-update-virtualenv` command. This command will initialize the virtualenv and install all packages needed by plugins shipped with syslog-ng itself from the Python Package Index (PyPI). If you need any additional Python libraries for your local scripts, you can install those using the `pip` command located in the virtualenv's bin directory; module='requests'
Traceback (most recent call last):
  File "/usr/local/etc/syslog-ng.conf{python-global-code:25}", line 3, in <module>
    import syslogng.modules.hypr
  File "/usr/local/lib/syslog-ng/python/syslogng/modules/hypr/__init__.py", line 48, in <module>
    import requests
ModuleNotFoundError: No module named 'requests'
Error parsing python, Error processing global python block in /usr/local/lib/syslog-ng/python/syslogng/modules/hypr/scl/hypr.conf:25:1-25:7:
20      # OpenSSL libraries as published by the OpenSSL project. See the file
21      # COPYING for details.
22      #
23      #############################################################################
24      
25----> python {
25----> ^^^^^^
26      
27      import syslogng.modules.hypr
28      
29      syslogng.modules.hypr.register_hypr_config_generator()
30      

Included from /usr/local/share/syslog-ng/include/scl/python/python-modules.conf:25:1-25:1:

Included from /usr/local/etc/scl.conf:31:1-31:1:
26      #
27      
28      @module appmodel
29      
30      @include 'scl/*/*.conf'
31----> 
31----> ^
32      @define java-module-dir "`module-install-dir`/java-modules"

Included from /usr/local/etc/syslog-ng.conf:3:1-3:1:
1       @version:4.2
2       @include "scl.conf"
3-----> 
3-----> ^
4       #
5       # This sample configuration file is essentially equilivent to the stock
6       # FreeBSD /etc/syslog.conf file.
7       #
8       
```

**Comment by bazsi** [comment]
We might defer the import of requests in the hypr case, that could
potentially be imported only if the generator is invoked.

Can you pls try what happens if the import requests is moved to the
function body in hypr?

**Pull request merged**